hudi-common/src/main/java/org/apache/hudi/common/table/read/UpdateProcessor.java

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * Interface used within the {@link HoodieFileGroupReader<T>} for processing updates to records in Merge-on-Read tables.
+ * Note that the updates are always relative to the base file's current state.
+ * @param <T> the engine specific record type
+ */
+public interface UpdateProcessor<T> {
+  /**
+   * Processes the update to the record. If the update should not be returned to the caller, the method should return null.
+   * @param recordKey the key of the record being updated
+   * @param previousRecord the previous version of the record, or null if there is no previous value
+   * @param currentRecord the current version of the record after merging with the existing record, if any exists
+   * @param isDelete a flag indicating whether the merge resulted in a delete operation
+   * @return the processed record, or null if the record should not be returned to the caller
+   */
+  T processUpdate(String recordKey, T previousRecord, T currentRecord, boolean isDelete);
+
+  static <T> UpdateProcessor<T> create(HoodieReadStats readStats, HoodieReaderContext<T> readerContext,
+                                       boolean emitDeletes, Option<BaseFileUpdateCallback> updateCallback) {
+    UpdateProcessor<T> handler = new StandardUpdateProcessor<>(readStats, readerContext, emitDeletes);
+    if (updateCallback.isPresent()) {
+      return new CallbackProcessor<>(updateCallback.get(), handler, readerContext);
+    }
+    return handler;
+  }
+
+  /**
+   * A standard update processor that increments the read stats and returns the record if applicable.
+   * @param <T> the engine specific record type
+   */
+  class StandardUpdateProcessor<T> implements UpdateProcessor<T> {
+    protected final HoodieReadStats readStats;
+    protected final HoodieReaderContext<T> readerContext;
+    protected final boolean emitDeletes;
+
+    public StandardUpdateProcessor(HoodieReadStats readStats, HoodieReaderContext<T> readerContext,
+                                   boolean emitDeletes) {
+      this.readStats = readStats;
+      this.readerContext = readerContext;
+      this.emitDeletes = emitDeletes;
+    }
+
+    @Override
+    public T processUpdate(String recordKey, T previousRecord, T currentRecord, boolean isDelete) {
+      if (isDelete) {
+        readStats.incrementNumDeletes();
+        if (emitDeletes) {
+          return readerContext.getDeleteRow(currentRecord, recordKey);
+        }
+        return null;
+      } else {
+        if (previousRecord != null) {
+          readStats.incrementNumUpdates();
+        } else {
+          readStats.incrementNumInserts();
+        }
+        return readerContext.seal(currentRecord);
+      }
+    }
+  }
+
+  /**
+   * A processor that wraps the standard update processor and invokes a customizable callback for each update.
+   * @param <T> the engine specific record type
+   */
+  class CallbackProcessor<T> implements UpdateProcessor<T> {
+    private final BaseFileUpdateCallback callback;
```

Comment:
If we want to avoid another specific interface for this callback, we can also make a more generic `TriConsumer` to take its place.